src/compiler/rust.rs

```diff
@@ -317,61 +317,193 @@ impl<T> Compiler<T> for Rust
+        let ArgLinkLibrary { kind, name } = self;
+        make_os_string!(kind, "=", name)
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+struct ArgLinkPath {
+    kind: String,
+    path: PathBuf,
+}
+impl FromArg for ArgLinkPath {
+    fn process(arg: OsString) -> ArgParseResult<Self> {
+        let (kind, path) = match split_os_string_arg(arg, "=")? {
+            (kind, Some(path)) => (kind, path),
+            // If no kind is specified, the path is used to search for all kinds
+            (path, None) => ("all".to_owned(), path),
+        };
+        Ok(ArgLinkPath { kind: kind, path: path.into() })
+    }
+}
+impl IntoArg for ArgLinkPath {
+    fn into_os_string(self) -> OsString {
+        let ArgLinkPath { kind, path } = self;
+        make_os_string!(kind, "=", path)
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+struct ArgCodegen {
+    opt: String,
+    value: Option<String>,
+}
+impl FromArg for ArgCodegen {
+    fn process(arg: OsString) -> ArgParseResult<Self> {
+        let (opt, value) = split_os_string_arg(arg, "=")?;
+        Ok(ArgCodegen { opt, value })
+    }
+}
+impl IntoArg for ArgCodegen {
+    fn into_os_string(self) -> OsString {
+        let ArgCodegen { opt, value } = self;
+        if let Some(value) = value {
+            make_os_string!(opt, "=", value)
+        } else {
+            make_os_string!(opt)
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+struct ArgExtern {
+    name: String,
+    path: PathBuf,
+}
+impl FromArg for ArgExtern {
+    fn process(arg: OsString) -> ArgParseResult<Self> {
+        if let (name, Some(path)) = split_os_string_arg(arg, "=")? {
+            Ok(ArgExtern { name, path: path.into() })
+        } else {
+            Err(ArgParseError::Other("no path for extern"))
+        }
+    }
+}
+impl IntoArg for ArgExtern {
+    fn into_os_string(self) -> OsString {
+        let ArgExtern { name, path } = self;
+        make_os_string!(name, "=", path)
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+enum ArgTarget {
+    Name(String),
+    Path(PathBuf),
+    Unsure(OsString),
+}
+impl FromArg for ArgTarget {
+    fn process(arg: OsString) -> ArgParseResult<Self> {
+        // Is it obviously a json file path?
+        if Path::new(&arg).extension().map(|ext| ext == "json").unwrap_or(false) {
+            return Ok(ArgTarget::Path(arg.into()))
+        }
+        // Time for clever detection - if we append .json (even if it's clearly
+        // a directory, i.e. resulting in /my/dir/.json), does the path exist?
+        let mut path = arg.clone();
+        path.push(".json");
+        if Path::new(&path).is_file() {
+            // Unfortunately, we're now not sure what will happen without having
+            // a list of all the built-in targets handy, as they don't get .json
+            // auto-added for target json discovery
+            return Ok(ArgTarget::Unsure(arg))
+        }
+        // The file doesn't exist so it can't be a path, safe to assume it's a name
```

Comment:
I wonder if we'd be better off running `rustc --print=target-list` in the initial compiler detection and matching against that?